python/tskit/tables.py

```diff
@@ -3991,6 +3991,96 @@ def trim(self, record_provenance=True):
+        """
+        self.edges.left += value
+        self.edges.right += value
+        self.sites.position += value
+        if sequence_length is None:
+            self.sequence_length += value
+        else:
+            self.sequence_length = sequence_length
+
+        if record_provenance:
+            parameters = {
+                "command": "shift",
+                "value": value,
+                "sequence_length": sequence_length,
+            }
+            self.provenances.add_row(
+                record=json.dumps(provenance.get_provenance_dict(parameters))
+            )
+
+    def concatenate(
+        self, other, *, node_mapping=None, record_provenance=True, **kwargs
+    ):
+        """
+        Concatenate another table collection to the right of this one. This
+        {meth}`shift`s the other table coordinate rightwards, then calls
+        {meth}`union` with ``check_shared_equality=False`` and the provided
+        ``node_mapping``. If no node mapping is given, the two table
+        collections must have the same number of samples, and those are treated
+        (in numerical order) as shared between the two table collections.
+        This is identical to :meth:`TreeSequence.concatenate` but
+        acts *in place* to alter the data in this :class:`TableCollection`.
+
+        .. note::
+            To add gaps between the concatenated tables, use :meth:`shift` before
+            concatenating; to remove gaps, use :meth:`trim`.
+
+        :param TableCollection other: The other table collection to add to the right
+            of this one.
+        :param list node_mapping: An array of integers of the same length as the number
+            of nodes in ``other``, where the _k_'th element gives the id of the node in
+            the current table collection corresponding to node _k_ in the other table
+            collection (see {meth}`union`). If None (default), only the sample nodes
+            between the two node tables, in numerical order, are mapped to each other.
+        :param bool record_provenance: If True (default), record details of this call to
+            ``concatenate`` in the returned tree sequence's provenance information
+            (Default: True).
+        :param \\**kwargs: Additional keyword arguments to pass to {meth}`union`
+            (e.g. ``add_populations``).
+        """
+        if node_mapping is None:
+            samples = np.where(self.nodes.flags & tskit.NODE_IS_SAMPLE)[0]
+            other_samples = np.where(other.nodes.flags & tskit.NODE_IS_SAMPLE)[0]
+            if len(other_samples) != len(samples):
+                raise ValueError(
+                    "each `other` must have the same number of samples as `self`"
+                )
+            node_mapping = np.full(other.nodes.num_rows, tskit.NULL, dtype=np.int32)
+            node_mapping[other_samples] = samples
+        other.shift(self.sequence_length, record_provenance=False)
```

Comment:
Ah, I hadn't clocked that this formulation would required that we alter the parameter. That's pretty nasty and surprising, and I don't thinks we should do it.

Reply:
Oh, that's a very good point. I wonder if we can avoid that somehow. E.g. could we add the other to the *left* of self? That way we only need to shift the coordinate system of `self` every time.

Alternatively, we could temporarily shift the coordinates of `self` so that they are negative, and end at zero, then union with `other`, then shift the whole lot rightwards again. I think that avoids modifying `other`, but at the expense of calling `shift` twice rather than once.

Reply:
Let's not bother and just do it for tree sequences instead.

Reply:
I just checked. We can't `union` a table collection with negative positions. But we could add to the left easily enough, which doesn't alter `other`. The naming would probably need changing. E.g. the TableCollection method could be `prepend`, and then we could easily use that in a TreeSequence.concatenate() function:

```python
def concatenate(self, other):
    tables = other.dump_tables()
    tables.prepend(self.tables)
    return tables.tree_sequence()
```

Reply:
Please, let's not bother - the TreeSequenced version is all we need and has the required semantics.